Add additional statistics for WindowOperator in Explain Analyze continued

maciejgrzybek opened the pull request:
Statistics are low level, therefore `explain analyze verbose` mode was added (syntax as in PostgreSQL; https://www.postgresql.org/docs/9.6/static/sql-explain.html).

Example output:

NON-VERBOSE

```sql
         - Window[partition by (shipmode)][$hashvalue] => [quantity:double, shipmode:varchar(10), $hashvalue:bigint, sum:double]
                 Cost: 92.60%, Output: 60175 rows (2.08MB)
                 Input avg.: 3760.94 lines, Input std.dev.: 139.24%
                 sum := sum("quantity")
```

---

VERBOSE:
```sql
         - Window[partition by (shipmode)][$hashvalue] => [quantity:double, shipmode:varchar(10), $hashvalue:bigint, sum:double]
                 Cost: 94.33%, Output: 60175 rows (2.08MB)
                 Input avg.: 3760.94 lines, Input std.dev.: 139.24%
                 Active Drivers: [ 6 / 16 ]
                 Index size in bytes: { count: 23, total: 5238332, min: 0, p05: 0, p25: 0, p50: 316238, p75: 329535, p95: 648366, max: 648366 }
                 Rows in partitions per index: { count: 23, total: 60175, min: 0, p05: 0, p25: 0, p50: 0, p75: 8491, p95: 8669, max: 8710 }
                 Rows in partitions per driver: { count: 16, total: 60175, min: 0, p05: 0, p25: 0, p50: 0, p75: 8616, p95: 17192, max: 17192 }
                 sum := sum("quantity")
```

---

- `Active drivers` is a number of operators which had any non-empty data set to process to number of all available drivers.
- `Index size in bytes` is a distribution of the size of index among all the drivers. This may help identify problems when some indexes were bigger than others (data skew / unnecessarily built indexes [possibility to reuse maybe?]).
- `Rows in partitions per index` - how big were partitions per index (index may be built for more than one partition).
- `Rows in partitions per driver` tells us how many rows per partition each driver had to handle (this may help to identify data skew but differently than `index size in bytes` because it's possible to have evenly distributed size of indexes while rows per partition are not, e.g. one big index with single partition and another index with multiple small partitions).

They are very low level statistics. Even advanced Presto user may not understand them but they would be helpful to debug the implementation of WF operator (by us, developers) and identify problems visible at customer's site.
`Explain analyze verbose` simply provides a shortcut to extra information exposed in JSON.

---

Supersedes https://github.com/prestodb/presto/pull/7382.

maciejgrzybek commented:
Information about the distribution of the data will be added to the analyze of `Exchange` nodes, as per discussion with @martint.
This is because the information about data distribution could be used not only in WF context but also in JOINS or aggregates (simply any place where exchange happens).

maciejgrzybek commented:
@sopel39, @martint: thoughts?

sopel39 commented:
Please update `presto-docs/src/main/sphinx/sql/explain-analyze.rst` with more depth explanation what are the new stats and verbose option

sopel39 commented:
> Active drivers is a number of operators

This isn't precise. Do you mean the number of drivers that contain operator for a given plan node? There might be multiple operators for a plan node, which are not in the same drivers (e.g: `JoinNode`). I guess this info is window function specific. Additionally, can window function operators be in source stage (without any reshuffle)? In such case number of drivers will be equal to number of processed splits. Is the purpose of this stat to recognize skewness? If so, it seems that other stats duplicate it (e.g: avg and std.dev of input rows, partition stats). On the other hand, if we want to report empty splits, maybe it shouldn't be coupled with window function stats.

```
                 Rows in partitions per index: { count: 23, total: 60175, min: 0, p05: 0, p25: 0, p50: 0, p75: 8491, p95: 8669, max: 8710 }
                 Rows in partitions per driver: { count: 16, total: 60175, min: 0, p05: 0, p25: 0, p50: 0, p75: 8616, p95: 17192, max: 17192 }
```

Why simple average and std.dev isn't enough? If the goal is to recognize problem with data (skewness) than those two numbers might be enough and actually provide clearer information about problems with data. With percentiles you have to check if the numbers increment more or less linearly while high std.dev provides you with good indication that there is problem with skewness. In both cases you have to do deep dive to check what is the actual underlying issue with data. Maybe report both?

> Information about the distribution of the data will be added to the analyze of Exchange nodes

Do you mean avg and std.dev or exact percentiles?

sopel39 commented:
> is a number of operators which had any non-empty

What is so special about empty/non-empty separation? What if all operators process one row, but one of the operators process all other rows. In such case you would get `total drivers: n out of n`, which could be misleading that data is parallelized.

maciejgrzybek commented:
@sopel39: indeed, this info is very much WF-specific.
Also, I showed the example output of the operator (with distribution) to @martint and he said it looks good to him.
This does not have to be pretty. We want it to be VERBOSE, as would name of the mode suggest, so we could e.g. rebuild customers case based solely on the `explain analyze verbose` output, without disclosing theirs data sets.

Also, you make a basic mistake here by assuming it's for _average user_:
> Average user might find it difficult to interpret percentiles.

It's not meant for average user. It's not even meant for DBA/expert user. It's for developers. That's why it's hidden in `verbose` mode.

> Additionally, can window function operators be in source stage (without any reshuffle)?
Yes, this may happen when data is preSorted/prePartitioned. It happens when connector provides some info about ordering. Although, then WF works in different mode.
I think it's okay to have those stats less helpful in such a case, especially given it's rare case to have WF working in stream-mode. Currently such a case happens only for MySQL deployment at FB.

> Do you mean avg and std.dev or exact percentiles?

For exchanges we'd have avg and sd because it will be customer facing feature (non-verbose mode).

maciejgrzybek commented:
@sopel39: Added description in docs in 348eb79.

Rebased on master and resolved conflicts caused by @ArturGajowy's refactor.

@martint: can you comment or dispatch to someone to have a look on that?

maciejgrzybek commented:
@raghavsethi let me know when you have time for this PR, so I don't have to unnecessarily rebase and resolve conflicts.

raghavsethi commented:
Going to discuss the explain analyze-related changes with @martint tomorrow. Reviewing the other parts now. Please rebase your PR when you get a chance.

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/operator/OperatorInfo.java:
```diff
@@ -14,6 +14,7 @@
 package com.facebook.presto.operator;
 
 import com.facebook.presto.operator.PartitionedOutputOperator.PartitionedOutputInfo;
+import com.facebook.presto.operator.WindowOperator.WindowInfo;
```
Commit message doesn't need to discuss explain analyze at all, they're independent.

maciejgrzybek replied:
Removed that line from the summary.

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:
```diff
@@ -480,4 +501,143 @@ private static int findGroupEnd(PagesIndex pagesIndex, PagesHashStrategy pagesHa
+        private final long sizeInBytes;
+        private final List<Integer> partitions;
+
+        public IndexInfo(
```
`@JsonCreator`?

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:
```diff
@@ -480,4 +501,143 @@ private static int findGroupEnd(PagesIndex pagesIndex, PagesHashStrategy pagesHa
+    }
+
+    @ThreadSafe
+    public static class SingleDriverWindowInfoBuilder
```
DriverWindowInfoBuilder

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:
```diff
@@ -480,4 +501,143 @@ private static int findGroupEnd(PagesIndex pagesIndex, PagesHashStrategy pagesHa
+    }
+
+    @Immutable
+    public static class SingleDriverWindowInfo
```
DriverWindowInfo

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:
```diff
@@ -480,4 +501,143 @@ private static int findGroupEnd(PagesIndex pagesIndex, PagesHashStrategy pagesHa
+    @ThreadSafe
+    public static class SingleDriverWindowInfoBuilder
+    {
+        private List<IndexInfo> indexInfos = new ArrayList<>();
```
Why not just make this a builder?

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:
```diff
@@ -480,4 +501,143 @@ private static int findGroupEnd(PagesIndex pagesIndex, PagesHashStrategy pagesHa
+        {
+            this.rowsNumber = rowsNumber;
+            this.sizeInBytes = sizeInBytes;
+            this.partitions = ImmutableList.builder();
```
Inline this

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:
```diff
@@ -192,6 +192,7 @@ public OperatorFactory duplicate()
     private final PageBuilder pageBuilder;
 
     private final SingleDriverWindowInfoBuilder windowInfo;
+    private SingleDriverWindowInfo singleDriverWindowInfo;
```
Merge this with the commit that introduced it. Also, I don't see what you're referring to in HashAggregationOperator could you be more specific?

maciejgrzybek replied:
Removed reference to `HashAggregationOperator` and added one line of explanation (0af2be44bc78a40530c28c95084eabb6c84597f0).

maciejgrzybek commented:
Comments addressed, @raghavsethi.

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:
```diff
@@ -480,4 +500,141 @@ private static int findGroupEnd(PagesIndex pagesIndex, PagesHashStrategy pagesHa
+        @JsonCreator
+        public WindowInfo(@JsonProperty("windowInfos") List<DriverWindowInfo> windowInfos)
+        {
+            this.windowInfos = ImmutableList.copyOf(windowInfos);
```
requireNonNull

maciejgrzybek replied:
`ImmutableList.copyOf` internally check for non-nullness, so it's not necessary to add extra `requireNonNull` here.

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStats.java:
```diff
@@ -41,6 +42,7 @@
 
     private final Map<String, OperatorInputStats> operatorInputStats;
     private final Map<String, OperatorHashCollisionsStats> operatorHashCollisionsStats;
+    private final Optional<WindowNodeStats> windowNodeStats;
```
Not a fan of this direction - this is making PlanNodeStats more and more of a union structure. OperatorHashCollisionsStats is a little bit better because it applies to the couple of operators that build a hashmap.

Can we somehow move this in a more general direction? We could use inheritance, or maybe we could add printing support to OperatorInfo?

Need to discuss this with @martint, but I would like us to at least consider other ideas rather than making this class more complex.

maciejgrzybek replied:
Make sense but regardless what you agree on, I claim such a change should be a separate PR (to apply new schema to all stats at once).

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStats.java:
```diff
@@ -50,7 +52,8 @@
             Map<String, OperatorInputStats> operatorInputStats,
-            Map<String, OperatorHashCollisionsStats> operatorHashCollisionsStats)
+            Map<String, OperatorHashCollisionsStats> operatorHashCollisionsStats,
+            Optional<WindowNodeStats> windowNodeStats)
```
In any case, WindowNodeStats is the wrong abstraction - this class is about operators. Introducing a plan node-related member variable is weird.

maciejgrzybek replied:
Addressed in f094637 (kept as a fixup, so you could continue with remaining part of the review and then confirming that the addressed part looks good too).

raghavsethi commented:
I discussed this with @martint and we decided that we should probably pull all merging/formatting of operator infos to explain-ready strings in a separate class. Even a method that with a switch would work. We can move ahead with the PR for now, but I would like to see comments that make it clear that this flow is deprecated. Please fixup, squash and rebase, and I will continue to review.

maciejgrzybek commented:
@raghavsethi does 7b3bd72 seem like a good enough indication to rewrite such a class?

What do you think of an idea of having a base class of `PlanNodeStats` which would be extended for each operator (when extra info has to be provided) with appropriate `toString` (or some custom method taking into consideration indentation in multi-line outputs) method formatting output as needed for `PlanPrinter`? Then `PlanNodeStatsSummarizer` would return a list of `? extends PlanNodeStats`. `PlanPrinter` would simply go through the list of `PlanNodeStats` (presumably presorted by `PlanNodeId`) and call `toString()` method for each of the elements.

How does it sound?

If that's okay, then please merge this PR and I start refactoring right away. I want to know that we have a stable baseline first, though.

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:
```diff
@@ -517,4 +537,141 @@ else if (!previousPairsWereEqual) {
+    }
+
+    @ThreadSafe
+    public static class DriverWindowInfoBuilder
```
Why is this public?

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStatsSummarizer.java:
```diff
@@ -118,6 +126,12 @@ private PlanNodeStatsSummarizer() {}
                 }
 
+                // The only statistics we have for Window Functions are very low level, thus displayed only in VERBOSE mode
+                if (operatorStats.getInfo() instanceof WindowOperator.WindowInfo && verbose) {
```
This is also super weird - verbose should control what stats are *displayed*, not what stats are available. What if there were actually some stats about WindowOperator in both categories? It seems like verbosity should only be a factor during printing, and not otherwise. Please change the first commit to reflect this.

maciejgrzybek replied:
Well, the idea was to save merging and collecting unnecessary data if we're not going to display it anyway. Therefore the comment in the code above. If there will be any stats worth displaying for WFs in non-verbose mode, we'd need to rewrite it (remove the `if verbose` bit). If you don't like it, I can rewrite it now and collect stats every time and never display them (for now).

raghavsethi replied:
Merging is not really going to be very CPU intensive, and the data has already been transferred. We should make the change to collect everything and display according to display-specific settings.

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:
```diff
@@ -429,6 +434,25 @@ private void printDistributions(int indent, PlanNodeStats nodeStats)
         return ImmutableMap.of();
     }
 
+    private void printWindowOperatorStats(int indent, WindowOperatorStats stats)
```
i.e. the verbosity should likely be a parameter to this function.

maciejgrzybek replied:
Done.

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/WindowOperatorStats.java:
```diff
@@ -0,0 +1,131 @@
+        this.totalDrivers = totalDrivers;
+    }
+
+    public static Optional<WindowOperatorStats> merge(Optional<WindowOperatorStats> first, Optional<WindowOperatorStats> second)
```
Discussed with @martint, he is OK with moving Mergeable to the util package and using that instead.

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:
```diff
@@ -193,6 +193,7 @@ public OperatorFactory duplicate()
     private final PageBuilder pageBuilder;
 
     private final DriverWindowInfoBuilder windowInfo;
+    private DriverWindowInfo driverWindowInfo;
```
Change WindowInfo to materialize on operator close

This minimizes unnecessary data transfer and computation.

maciejgrzybek replied:
This already happens.
Expensive part (building `DriverWindowInfo`) happens on operator close only.
See b570455.
Until the operator closes, it produces empty `WindowInfo` which is wrapper around `List<DriverWindowInfo>`.
If you don't like it (i.e. claim it's too expensive to create an empty object with `new`, which I doubt is a problem in Java), I can return `null` there or static empty `WindowInfo` instead.

raghavsethi replied:
Sorry, I just wanted the commit message changed :)

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStats.java:
```diff
@@ -30,6 +30,7 @@
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 import static java.util.stream.Collectors.toMap;
 
+// TODO: break into into operator-specific stats classes instead of having a big union-class aggregating all stats together
```
Mark as @Deprecated

maciejgrzybek commented:
@raghavsethi: Rebased onto master. I extracted `WindowInfo` to separate class with helper classes as its subclasses.
f3cc763 moves `verbose` flag to the display method, as you requested. I kept it as a separate commit so it's easier to review only the diff.

maciejgrzybek commented:
We have discovered problem with previous version of patch we rolled into production. If data was partitioned by very selective column (low cardinality of each partition), we were producing long statistics. This caused problem with too big http buffer on worker node.
Therefore, I decided to pre-accumulate values before propagated to the coordinator.
So now we support only standard deviation in output instead of full distribution in order to lower the statistics size.
Currently, the statistics size is constant, regardless of partitioning scheme or any feature of the data.
The only "moving" part is that `WindowInfo` holds a *list* of `DriverWindowInfo`. However, this list contains one element per task, so currently it's not possible to have a worker with more than one element in that list, thus it's safe.

---

Currently, the output looks like:
```
presto:tiny> explain analyze verbose select sum(discount) over(partition by orderkey) from lineitem;
                                                                    Query Plan
--------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 1 [tpch:orders:15000]
     CPU: 332.42ms, Input: 60175 rows (1.55MB), Output: 60175 rows (528.88kB)
     Output layout: [sum]
     Output partitioning: SINGLE []
     - Project[] => [sum:double]
             Cost: {rows: ?, bytes: ?}
             CPU fraction: 4.02%, Input: 60175 rows (2.07MB), Output: 60175 rows (528.88kB), Filtered: 0.00%
             Input avg.: 15043.75 lines, Input std.dev.: 0.46%
         - Window[partition by (<orderkey>)][$hashvalue] => [orderkey:bigint, discount:double, $hashvalue:bigint, sum:double]
                 Cost: {rows: ?, bytes: ?}
                 CPU fraction: 58.29%, Output: 60175 rows (2.07MB)
                 Input avg.: 15043.75 lines, Input std.dev.: 0.46%
                 Active Drivers: [ 4 / 4 ]
                 Index size: std.dev.: 13069.38 bytes , 2.12 rows
                 Index count per driver: std.dev.: 0.00
                 Rows per driver: std.dev.: 69.06
                 Size of partition: std.dev.: 0.00
                 sum := sum("discount")
             - ScanProject[table = tpch:tpch:lineitem:sf0.01, originalConstraint = true] => [orderkey:bigint, discount:double, $hashvalue:bigint]
                     Cost: {rows: 60000, bytes: ?}/{rows: 60000, bytes: ?}
                     CPU fraction: 37.69%, Input: 60175 rows (0B), Output: 60175 rows (1.55MB), Filtered: 0.00%
                     Input avg.: 15043.75 lines, Input std.dev.: 0.46%
                     $hashvalue := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey"), 0))
                     orderkey := tpch:orderkey
                     discount := tpch:discount


(1 row)
```

Explanation:
`Active Drivers: [ x / y ]` -  *x* is a number of operators which had any non-empty data set to process, *y* stands for a number of all available drivers.
` Index size: std.dev.: x bytes , y rows` - how big were differences between sizes of indexes among drivers
`Index count per driver: std.dev.: x` - how many times we had to rebuild the index among drivers
`Rows per driver: std.dev.: x` - measure of rows count dispersion among drivers
`Size of partition: std.dev.: x` - variability of the partition size

---

I think you'd like this approach better, @sopel39.
Please skim through it if the math looks good to you, as well as stats still seem reasonable to you.

sopel39 commented:
Thanks. I will take a look after I'm done with join stats code

maciejgrzybek commented:
@sopel39 We probably want to include this in the release, so we could better understand customer problems in production. Please review or delegate to someone else.

sopel39 commented:
I will review it now

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/WindowInfo.java:
```diff
@@ -0,0 +1,252 @@
+    }
+
+    @ThreadSafe
+    static class DriverWindowInfoBuilder
```
Is it used by single driver? If so, why this needs to be thread safe?

For instance, `com.facebook.presto.operator.HashCollisionsCounter` is not thread safe (but I think it might be a bug, counters inside should be atomic)

maciejgrzybek replied:
There's one `DriverWindowInfoBuilder` per `WindowOperator`, so yes, it's safe because accessed by single driver only.
But before `Prevent providing WindowInfo before the operator finishes` commit, `DriverWindowInfoBuilder` was materialized (`build()` method) in `getWindowInfo()` method, which is called by stats collectors (other thread than operator itself).
Here's a fixup to the `Prevent providing WindowInfo before the operator finishes` commit, removing `synchronized` from `DriverWindowInfoBuilder`: 30d3318eeda7aa35f9566c83d7266def92e249f3.

sopel39 replied:
Why not make `DriverWindowInfoBuilder` a supplier of `DriverWindowInfo` similar as `HashCollisionsCounter`? This way we could get continuous updates on window operator status. We could display it then in GUI or in (future) continuous explain analyze

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/WindowInfo.java:
```diff
@@ -0,0 +1,252 @@
+        public synchronized void addIndex(PagesIndex index)
+        {
+            if (currentIndexInfoBuilder != null) {
+                Optional<IndexInfo> indexInfo = currentIndexInfoBuilder.build();
```
use `ifPresent`:
```
currentIndexInfoBuilder.build().ifPresent(indexInfosBuilder::add)
```

maciejgrzybek replied:
Fixed in 40c764c4ed0198902338f58e8d2e463b3f870e0b.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/WindowInfo.java:
```diff
@@ -0,0 +1,252 @@
+        public synchronized DriverWindowInfo build()
+        {
+            if (currentIndexInfoBuilder != null) {
+                Optional<IndexInfo> indexInfo = currentIndexInfoBuilder.build();
```
you can extract this logic of finishing current builder info separate method and use it here and in `addIndex`

maciejgrzybek replied:
While in `build()` method, we have to nullify `currentIndexInfoBuilder` in addition to adding index info to the builder.
I don't think it brings any more good if I extract this two lines to a separate method:
```java
                Optional<IndexInfo> indexInfo = currentIndexInfoBuilder.build();
                indexInfo.ifPresent(indexInfosBuilder::add);
```
given the `if` cannot be extracted because I need to nullify the `currentIndexInfoBuilder` in `build()` only.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/WindowInfo.java:
```diff
@@ -0,0 +1,252 @@
+                    .sum();
+            double averageIndexSize = indexInfos.stream()
+                    .mapToLong(index -> index.getSizeInBytes())
+                    .average().orElse(Double.NaN);
```
`average()` result should be there since `indexInfos` cannot be empty at this point.

maciejgrzybek replied:
Agreed, that's why `NaN` will never occur but `LongStream::average` returns `OptionalLong` while I want regular `double`, that's why `orElse` which will never return the "else-value".

sopel39 replied:
Then use `getAsDouble` because it will throw an exception if the value is not present (which is unexpected).

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/WindowInfo.java:
```diff
@@ -0,0 +1,252 @@
+            if (partitions.size() == 0) {
+                return Optional.empty();
+            }
+            double avgSize = partitions.stream().mapToLong(Integer::longValue).average().orElse(Double.NaN);
```
ditto, `average()` cannot return `empty()`

maciejgrzybek replied:
ditto

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/WindowInfo.java:
```diff
@@ -0,0 +1,252 @@
+    {
+        private final long totalRowsCount;
+        private final long sizeInBytes;
+        private final double sumSquaredDifferences; // sum of (partitionSize - averagePartitionSize)^2 for each partition
```
make more meaningful name as in: https://github.com/prestodb/presto/pull/7661/commits/afcacf6ddb515141c43c6b2ff7f209f9cee9bfda#diff-44e86b0fa3217812de1a11e39ecfeaabR123

maciejgrzybek replied:
Isn't it the same level of explanation here and the link you pasted? It looks pretty much the same to me. :)

sopel39 replied:
`sumSquaredDifferences` of what? https://github.com/prestodb/presto/commit/afcacf6ddb515141c43c6b2ff7f209f9cee9bfda#diff-44e86b0fa3217812de1a11e39ecfeaabR123 has this information

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:
```diff
@@ -441,6 +452,7 @@ private void sortPagesIndexIfNecessary()
                 startPosition = endPosition;
             }
         }
+        windowInfo.addIndex(pagesIndex);
```
It is a bit odd that `windowInfo.addIndex(pagesIndex)` is part of `sortPagesIndexIfNecessary()` (method name is unrelated). Maybe add another method, `finishPagesIndex()`:
```
void finishPagesIndex() {
  sortPagesIndexIfNecessary();
  windowInfo.addIndex(pagesIndex);
}
```
or at least add a comment

maciejgrzybek replied:
This is part of a logic of window functions execution.
But if you insist, here's fixup: 049f1a508180e6fb0f02991a8cb45695934d3048.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/WindowInfo.java:
```diff
@@ -119,7 +119,7 @@ public synchronized DriverWindowInfo build()
 
     @Immutable
-    private static class DriverWindowInfo
+    public static class DriverWindowInfo
```
this should be part of previous commit

maciejgrzybek replied:
I made it here because previously there was no need to make it public, until now.
I can move it there, if you feel it's important.

maciejgrzybek replied:
Commit amended.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStats.java:
```diff
@@ -30,6 +31,8 @@
 import static java.util.stream.Collectors.toMap;
 
+// TODO: break into operator-specific stats classes instead of having a big union-class aggregating all stats together
+@Deprecated
```
TODO doesn't make it deprecated since there is no alternative currently.

maciejgrzybek replied:
This was explicit request from @raghavsethi (https://github.com/prestodb/presto/pull/7661#discussion_r119244354).

maciejgrzybek commented on presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:
```diff
@@ -1545,35 +1545,69 @@ public void testExplain()
+    public void testExplainVerbose()
+            throws Exception
+    {
+        assertStatement("EXPLAIN VERBOSE SELECT * FROM t",
```
It depends on the operator. Currently, it's meant only for WF info because it's too detailed and too low level for average (even experienced) user to understand, so we don't want to pollute regular `explain analyze` with WF details which are as of now - implementation details-related.

sopel39 replied:
right, but does `EXPLAIN VERBOSE` (not analyze) is any different than `EXPLAIN`?

maciejgrzybek replied:
Currently not but could. Although, parser allows it, after my change, so I added a test to express that.

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/operator/WindowInfo.java:
```diff
@@ -0,0 +1,252 @@
+public class WindowInfo
+        implements Mergeable<WindowInfo>, OperatorInfo
+{
+    private final List<DriverWindowInfo> windowInfos;
```
Info was too hard to send, which I described here: https://github.com/prestodb/presto/pull/7661#issuecomment-307498062 but it was due to the idea of displaying whole distribution instead of std dev only.
In order to display the distribution you need to carry the information over (it's not mergeable). In the same time, std dev can be pre-accumulated and delivered to coordinator for final display. Each driver have only one `DriverWindowInfo` in the `windowInfos` list. Although, I use a `List` to make it easily mergeable on the coordinator (coordinator can simply call `OperatorInfo.mergeWith()` as it does for any other mergeable operator).

maciejgrzybek replied:
So no, it's not possible to have more than sizeof `DriverWindowInfo` data to send from single operator instance.

sopel39 replied:
sure, but can `List<DriverWindowInfo>` be unbounded? Can `WindowOperator` be in the same pipeline as `TableScanNode` (I guess not)?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/WindowOperatorStats.java:
```diff
@@ -0,0 +1,167 @@
+                long numberOfIndexes = driverWindowInfo.getNumberOfIndexes();
+                long rowCount = driverWindowInfo.getTotalRowsCount();
+
+                partitionRowsSumSquaredDiffs += driverWindowInfo.getSumSquaredDifferencesSizeInPartition();
```
Shouldn't it (and https://github.com/prestodb/presto/pull/7661/commits/45a7f66a6e89d87bc29bddc4af911cf5b88073be#diff-cf5d7dbb1d37bd1638bd50937d1152f0R71) be weighted somehow by number of of rows that driver processed (similarly as hash collisions computations do)? Now an operator that processed 1% of rows will have the same impact on stats as operator that processed 99% of rows.

Currently, either driver processed some rows (`partitionRowsCount > 0`) or it didn't, but there is a lot of in-between states (driver processed 1 row, driver processed very little/little/huge number of rows). Why processed/non-processed is more special division?

maciejgrzybek replied:
`partitionRowsSumSquaredDiffs` is used to determine std dev of rows in partitions, so with the assumption that every `WindowOperator` must have at least one partition (and always full partitions, never fractions), it shouldn't matter how many rows it processed per driver, to properly answer the question "how dispersed are rows among partitions" IMO, correct me if I'm wrong, though.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/WindowOperatorStats.java:
```diff
@@ -0,0 +1,167 @@
+                .orElse(Double.NaN);
+
+        for (WindowInfo.DriverWindowInfo driverWindowInfo : info.getWindowInfos()) {
+            long partitionRowsCount = driverWindowInfo.getTotalRowsCount();
```
Why `windowInfo.getTotalRowsCount()` becomes `partitionRowsCount`? I understand window operator can have more than one partition as in https://github.com/prestodb/presto/pull/7661/commits/45a7f66a6e89d87bc29bddc4af911cf5b88073be#diff-cf5d7dbb1d37bd1638bd50937d1152f0R67

maciejgrzybek replied:
For the lack of better name, I used `partitionRowsCount` even though it represents `totalRowsCount` for that operator, i.e. an operator may have 2 partitions, so that number represents total number of rows from both partitions.
Although, there's a global counter `totalRowCount` (https://github.com/prestodb/presto/pull/7661/files#diff-cf5d7dbb1d37bd1638bd50937d1152f0R43) which is used to count total number of rows in all partitions among all operators. If you have an idea for a better non-clashing name, please share it. :)

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/WindowOperatorStats.java:
```diff
@@ -0,0 +1,167 @@
+            totalDrivers++;
+            if (partitionRowsCount > 0) {
+                long numberOfIndexes = driverWindowInfo.getNumberOfIndexes();
+                long rowCount = driverWindowInfo.getTotalRowsCount();
```
this is the same value as `partitionRowsCount`. Don't introduce new variable for it

maciejgrzybek replied:
You're right. Fixed in 9eb2fe6e4c438b1775e533904b811c2f473245e8.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java:
```diff
@@ -529,4 +530,11 @@ else if (!previousPairsWereEqual) {
+    public void close()
+            throws Exception
+    {
+        driverWindowInfo = Optional.of(windowInfo.build());
```
this will make window info available only after operator is finished. I think it would be better if we can have continuous window information similar as for hash collisions and others. This will enable as to display them continuously in GUI or `EXPLAIN ANALYZE`.

maciejgrzybek replied:
I see your point, it COULD be useful to see that a WF was going smoothly in the beginning of the execution but went south due to skew at some point, although currently all our use cases (analysis of customers performance problems) require info gathered in `explain analyze` in the end of the query. It's not worth to pollute network and use CPU to compute intermediate results we don't want, IMO.
If we ever end up with a desire to read intermediate stats for WFs, we can enable them then.

Also, AD the first statement, I'm not sure what value without additional instrumentation it gives us to know that WF was executing fine for 3 minutes and started having troubles for the remaining e.g. 7 minutes of its 10 minute-time execution.

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/WindowOperatorStats.java:
```diff
@@ -0,0 +1,167 @@
+                .filter(windowInfo -> windowInfo.getTotalRowsCount() > 0)
+                .mapToLong(windowInfo -> windowInfo.getNumberOfIndexes())
+                .average()
+                .orElse(Double.NaN);
```
Having zero indexes is not allowed state, that's why I wanted to express the wrong of stats collection by simply displaying "NaN" to the user, instead of masking the problem behind, simply displaying wrong number (e.g. `Math.pow(numberOfIndexes - averageNumberOfIndexes, 2)` would return `0` which is not so worrisome as `NaN`, at least at first glance (because after analyzing the output, you'd notice that zero is not legal value there too).

sopel39 replied:
> Having zero indexes is not allowed state

Why not do check state then? `NaN` might suggest missing stats (not wrong state).

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/WindowOperatorStats.java:
```diff
@@ -0,0 +1,167 @@
+                .filter(windowInfo -> windowInfo.getTotalRowsCount() > 0)
+                .mapToLong(windowInfo -> windowInfo.getTotalRowsCount())
+                .average()
+                .orElse(Double.NaN);
```
ditto - we shouldn't have any stats collected if there were no drivers executing `WindowOperator`, otherwise `WindowInfo.windowInfos` can never be empty.

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/WindowOperatorStats.java:
```diff
@@ -0,0 +1,167 @@
+
+    public static WindowOperatorStats create(WindowInfo info)
+    {
+        int activeDrivers = 0;
```
Empty `WindowInfo.windowInfos` is not valid state, see https://github.com/prestodb/presto/pull/7661/files#r124903954.

maciejgrzybek replied:
Added an assertion in 8ed7a1c8a34ed71592646da3936e147723520333.

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/WindowOperatorStats.java:
```diff
@@ -0,0 +1,167 @@
+        for (WindowInfo.DriverWindowInfo driverWindowInfo : info.getWindowInfos()) {
+            long partitionRowsCount = driverWindowInfo.getTotalRowsCount();
+            totalDrivers++;
+            if (partitionRowsCount > 0) {
```
We don't want to accumulate information about non-working drivers in our std dev. We want to have higher accuracy for working drivers because we may derive information about level of concurrency from `active/total drivers count`.
E.g. when there're only two drivers working but 14 is doing nothing, we want to know with high accuracy how well the data was distributed among those two working ones.
But if there's a situation when 2 drivers is taking 99% of work while remaining 14 is having 1 row to process each, then we're interested in such a skew because somehow we ended up with heavy skew which cannot be detected by comparing number of active to non active drivers.

sopel39 commented:
squash, I will do final review and merge. CC @martint

maciejgrzybek commented:
@sopel39: squashed and travis is finally green (had to restart to get rid of intermittent failures).
